### ISSUE TITLE
VerneMQ getting killed and using a lot of memory

### ISSUE THREAD

**Issue by ranvijayj** [erlio/vernemq#587]
### Environment

- VerneMQ Version: 1.2.1
- OS: Ubuntu 16.04
- Erlang/OTP version (if building from source): deb package

### Expected behavior

verneMQ should keep running and not use up the entire memory available
leveldb is also set to 20%

- Please describe the expected behaviour.

### Actual behaviour

```2018-01-22 07:45:33 =CRASH REPORT====
  crasher:
    initial call: riak_sysmon_filter:init/1
    pid: <0.346.8>
    registered_name: []
    exception exit: {{noproc,{gen_server,call,[kernel_safe_sup,{start_child,{timer_server,{timer,start_link,[]},permanent,1000,worker,[timer]}},infinity]}},[{gen_server,init_it,6,[{file,"gen_server.erl"},{line,352}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [riak_sysmon_sup,<0.135.0>]
    messages: []
    links: [<0.136.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 376
    stack_size: 27
    reductions: 2105
  neighbours:
2018-01-22 07:45:33 =SUPERVISOR REPORT====
     Supervisor: {local,riak_sysmon_sup}
     Context:    start_error
     Reason:     {noproc,{gen_server,call,[kernel_safe_sup,{start_child,{timer_server,{timer,start_link,[]},permanent,1000,worker,[timer]}},infinity]}}
     Offender:   [{pid,{restarting,<0.140.0>}},{id,riak_sysmon_filter},{mfargs,{riak_sysmon_filter,start_link,[]}},{restart_type,permanent},{shutdown,2000},{child_type,worker}]

2018-01-22 07:45:33 =SUPERVISOR REPORT====
     Supervisor: {local,riak_sysmon_sup}
     Context:    shutdown
     Reason:     reached_max_restart_intensity
     Offender:   [{pid,{restarting,<0.140.0>}},{id,riak_sysmon_filter},{mfargs,{riak_sysmon_filter,start_link,[]}},{restart_type,permanent},{shutdown,2000},{child_type,worker}]

```

And in error.log 

```2018-01-22 07:45:33.389 [error] <0.140.12> CRASH REPORT Process <0.140.12> with 0 neighbours exited with reason: no such process or port in call to gen_server:call(kernel_safe_sup, {start_child,{timer_server,{timer,start_link,[]},permanent,1000,worker,[timer]}}, infinity) in gen_server:init_it/6 line 352
2018-01-22 07:45:33.389 [error] <0.136.0> Supervisor riak_sysmon_sup had child riak_sysmon_filter started with riak_sysmon_filter:start_link() at {restarting,<0.140.0>} exit with reason no such process or port in call to gen_server:call(kernel_safe_sup, {start_child,{timer_server,{timer,start_link,[]},permanent,1000,worker,[timer]}}, infinity) in context start_error
2018-01-22 07:45:33.389 [error] <0.346.8> CRASH REPORT Process <0.346.8> with 0 neighbours exited with reason: no such process or port in call to gen_server:call(kernel_safe_sup, {start_child,{timer_server,{timer,start_link,[]},permanent,1000,worker,[timer]}}, infinity) in gen_server:init_it/6 line 352
2018-01-22 07:45:33.390 [error] <0.136.0> Supervisor riak_sysmon_sup had child riak_sysmon_filter started with riak_sysmon_filter:start_link() at {restarting,<0.140.0>} exit with reason no such process or port in call to gen_server:call(kernel_safe_sup, {start_child,{timer_server,{timer,start_link,[]},permanent,1000,worker,[timer]}}, infinity) in context start_error
2018-01-22 07:45:33.390 [error] <0.136.0> Supervisor riak_sysmon_sup had child riak_sysmon_filter started with riak_sysmon_filter:start_link() at {restarting,<0.140.0>} exit with reason reached_max_restart_intensity in context shutdown
```

- Please include minimal steps to reproduce the issue, if possible.

Just running verneMQ normally. Conf settings
https://gist.github.com/ranvijayj/f7301f70296b408d8d2b8c16675d6966

So, leveldb I set to 20 (%) memory in vernemq.conf.. How can I change buffer size for leveldb.. Shows a very high number in generated conf above ^^

The whole memory gets used up and `vernemq gets killed`. What can I change for less memory usage?

**Comment by ioolkos** [comment]
Hi @ranvijayj thanks for reporting.
How many clients do you like to connect? How big is your RAM memory? 
Most importantly: how big are your OS side TCP buffers? huge buffers will add up using a lot of memory.

also: you don't need 40k acceptors per listener! set that to 400 or 1k...

**Comment by ranvijayj** [comment]
We are running this on a c4.large for now. 3.75 GB ram.
There are a lot of clients connecting to the cluster. We have 3 nodes in verneMQ cluster to which the clients connect. 
One node has 8 GB ram which is working totally fine. The other 2 get killed. Need to contain VMQ to work within 3.75 GB

40k listeners because earlier it was default 100 and we saw rate of more than 5000/sec. So that number is high...

I now just decreased async threads from 64 to 16.. memory usage went down but then crashed again...
These are the logs which showed up now:https://gist.github.com/ranvijayj/b1a664739d9d302172dce385d0c44f88

OS settings

```net.ipv4.tcp_syncookies = 1
fs.file-max = 2097152
net.ipv4.ip_local_port_range = 10000 65000
net.core.rmem_max = 16777216
net.core.wmem_max = 16777216
net.core.rmem_default = 16777216
net.core.wmem_default = 16777216
net.core.optmem_max = 40960
net.core.netdev_max_backlog = 50000
net.ipv4.tcp_max_syn_backlog = 30000
net.ipv4.tcp_max_tw_buckets = 2000000
net.ipv4.tcp_tw_reuse = 1
net.ipv4.tcp_fin_timeout = 15
net.ipv4.tcp_slow_start_after_idle = 0
net.ipv4.udp_rmem_min = 8192
net.ipv4.udp_wmem_min = 8192
net.ipv4.conf.all.log_martians = 0
net.ipv4.tcp_rmem = 4096 87380 8388608
net.ipv4.tcp_wmem = 4096 65536 8388608
net.ipv4.tcp_window_scaling = 1
kernel.pid_max = 65536
net.core.somaxconn = 40960
net.ipv4.tcp_keepalive_time = 180
net.ipv4.tcp_keepalive_probes = 5
net.ipv4.tcp_keepalive_intvl = 15```

Are you asking about the above? ^^

@ioolkos

**Comment by larshesel** [comment]
Hi

If I remember correctly then the lines: 

```
net.ipv4.tcp_rmem = 4096 87380 8388608
net.ipv4.tcp_wmem = 4096 65536 8388608
```

Mean that the socket read buffer default is 87380 bytes and the write buffer is `65536` (the three values are read: min, default, max). This means if you start 10K(just picked a random number here) connections this will use 10K*(87380+65536) ~ 1.5 Gigabytes. Besides this, Erlang also has an internal buffer for each socket allocated which is the max of the write and read buffers. So in this case the memory usage would be sth like 10K*(2*87380+65536) ~ 2.3Gigabytes for 10K clients at a minimum, which would be a substantial amount of the available 3.5G.

**Comment by ranvijayj** [comment]
Okay. Let me experiment with these numbers again. Thanks for this wonderful explanation. I had been using these numbers with my older Mosquitto MQTT setup but never gave problems. I think the `listener.nr_of_acceptors` being high as 40k could be the issue right?

I think for VerneMQ connections we can keep the below at smaller numbers right?
net.ipv4.tcp_rmem
net.ipv4.tcp_wmem
Probably at 32KB max

When would `8388608` in the last column be used in the above setting? 

Also, how can I set `max erlang memory` any idea?

I also reduced erlang async threads in vernemq.conf count to 16 and memory usage is lower now..Default was 64. 

@larshesel 

Thanks a lot for your inputs....

**Comment by ranvijayj** [comment]
somehow failed 
```gen_server vmq_cluster_mon terminated with reason: no such process or port in call to [{gen,do_for_proc,2,[{file,[103,101,110,46,101,114,108]},{line,261}]},{gen_event,rpc,2,[{file,[103,101,110,95,101,118,101,110,116,46,101,114,108]},{line,197}]},{vmq_cluster_mon,terminate,2,[{file,[47,111,112,116,47,118,101,114,110,101,109,113,47,100,105,115,116,100,105,114,47,49,46,50,46,49,47,95,98,117,105,108,100,47,100,101,102,97,117,108,116,47,108,105,98,47,118,109,113,95,115,101,114,118,101,114,47,115,114,99,47,118,109,113,95,99,108,117,115,116,101,114,95,109,111,110,46,101,114,108]},{line,153}]},{gen_server,try_terminate,3,[{file,[103,101,110,95,115,101,114,118,101,114,46,101,114,108]},{line,629}]},{gen_server,terminate,7,[{file,[103,101,110,95,115,101,114,118,101,114,46,101,114,108]},{line,795}]},{proc_lib,init_p_do_apply,3,[{file,[112,114,111,99,95,108,105,98,46,101,114,108]},{line,247}]}]
2018-01-22 10:40:32.050 [error] <0.29952.6> CRASH REPORT Process vmq_cluster_mon with 0 neighbours exited with reason: no such process or port in call to gen_server:terminate/7 line 800
2018-01-22 10:40:32.050 [error] <0.29951.6> Supervisor vmq_cluster_node_sup had child vmq_cluster_mon started with vmq_cluster_mon:start_link() at <0.29952.6> exit with reason noproc in context child_terminated```

**Comment by ioolkos** [comment]
What were you doing at this point?

**Comment by ranvijayj** [comment]
@ioolkos I just changed the tcp mem option to 20000 the second column... I allowed anonymous on as well to try to reduce load..

One of the servers again got killed even though memory was not full

```2018-01-22 12:28:27.516 [error] <0.310.0>@vmq_queue_sup:start_queue:233 vmq_queue_sup can't start vmq_queue for {[],<<"user@3bf43fa95f5a6874d89f0fb1c88f62ff9a7902a4">>} due crash error:system_limit
2018-01-22 12:28:27.516 [error] emulator Too many processes

2018-01-22 12:28:27.518 [error] <0.316.0>@vmq_queue_sup:start_queue:233 vmq_queue_sup can't start vmq_queue for {[],<<"user@90dfda5a0ba232b1abda506f1aad0cabe043fb66">>} due crash error:system_limit
2018-01-22 12:28:27.518 [error] emulator Too many processes

2018-01-22 12:28:27.519 [error] <0.285.0>@vmq_queue_sup:start_queue:233 vmq_queue_sup can't start vmq_queue for {[],<<"user@116e82d10cf1cd2c273e4ba633dd62aebdbe98ad">>} due crash error:system_limit
2018-01-22 12:28:27.519 [error] emulator Too many processes

2018-01-22 12:28:27.521 [error] <0.303.0>@vmq_queue_sup:start_queue:233 vmq_queue_sup can't start vmq_queue for {[],<<"user@2b227d669f4e869e09da988e6db8f44f20234d65">>} due crash error:system_limit
2018-01-22 12:28:27.521 [error] emulator Too many processes

2018-01-22 12:28:27.525 [error] <0.287.0>@vmq_queue_sup:start_queue:233 vmq_queue_sup can't start vmq_queue for {[],<<"user@60990f98c3f5c308fdd8720daebba666fdafbbee">>} due crash error:system_limit
2018-01-22 12:28:27.525 [error] emulator Too many processes

2018-01-22 12:28:27.527 [error] error Ranch listener {{0,0,0,0},5555} terminated with reason: system_limit
2018-01-22 12:28:27.527 [error] emulator Too many processes

2018-01-22 12:28:27.530 [error] emulator Too many processes

2018-01-22 12:28:27.531 [error] <0.312.0>@vmq_queue_sup:start_queue:233 vmq_queue_sup can't start vmq_queue for {[],<<"user@da2c475879e2075dfaafe7a89bba5a5daa178aa0">>} due crash error:system_limit
2018-01-22 12:28:27.545 [error] emulator Too many processes
```

Any clue? @ioolkos 

Does verneMQ have memory leak issues generally?

**Comment by ioolkos** [comment]
Your running into a configured process limit, and then VerneMQ cannot setup more queues.
You can configure this via the conf file:
`erlang.process_limit = 120000` in the vernemq.conf file
This corresponds to the +P erlang emulator flag.

**Comment by ranvijayj** [comment]
It was already at 100000. All this is so weird. 
I am keeping this issue open. Will be checking some more issues. 

Thanks till now @ioolkos

**Comment by larshesel** [comment]
You write it's at 100000 = 100K, which is a lot less than the proposed value of 1.2M. Could you verify exactly what the value is?

**Comment by ranvijayj** [comment]
Yeah, I had changed it on one node to check memory leak.
Works after I made it 10M.

So this error has gone and memory utilization is a little low but still getting used a lot..

Final settings 
```net.ipv4.tcp_syncookies = 1
fs.file-max = 2097152
net.ipv4.ip_local_port_range = 10000 65000
net.core.rmem_max = 16777216
net.core.wmem_max = 16777216
net.core.rmem_default = 16777216
net.core.wmem_default = 16777216
net.core.optmem_max = 40960
net.core.netdev_max_backlog = 50000
net.ipv4.tcp_max_syn_backlog = 30000
net.ipv4.tcp_max_tw_buckets = 2000000
net.ipv4.tcp_tw_reuse = 1
net.ipv4.tcp_fin_timeout = 15
net.ipv4.tcp_slow_start_after_idle = 0
net.ipv4.udp_rmem_min = 8192
net.ipv4.udp_wmem_min = 8192
net.ipv4.conf.all.log_martians = 0
net.ipv4.tcp_rmem = 4096 20000 16777216
net.ipv4.tcp_wmem = 4096 20000 16777216
net.ipv4.tcp_window_scaling = 1
kernel.pid_max = 65536
net.core.somaxconn = 40960
net.ipv4.tcp_keepalive_time = 300
net.ipv4.tcp_keepalive_probes = 5
net.ipv4.tcp_keepalive_intvl = 15```

`listener.nr_of_acceptors = 40000`
and
`leveldb.maximum_memory.percent = 30`

I don't know if these will help control memory getting maxed out again. I would also try 
`listener.nr_of_acceptors = 10000` to check if that makes it any better as recommended above.

Do you recommend more servers with fewer listeners each under the load balancer?

At the very minute, I have 2 c4.large (3.75GB)  and 1 c4.xlarge (7.5 GB) all for now are (2.2 GB used)


Will be monitoring and seeing if I get any errors and share here.
@larshesel Thanks!

**Comment by larshesel** [comment]
So, you're using 40K acceptors and default socket sizes of 20K - that will use around 2.3 GB (according to my calculations above). I don't understand why you need so many acceptors as you won't be able to accept 40K new connections/second due to other factors (disk-io/replicating metadata state to other nodes etc). I would recommend no more than 1-5K acceptors.

I'll close this issue as there aren't really any bugs reported here. But feel free to continue the discussion - but also note that you're welcome to discuss issue like these on our slack channel: https://slack-invite.vernemq.com/

**Comment by ranvijayj** [comment]
Suddenly this started happening

```Error in process <0.15651.1> on node 'VerneMQ@XXXX' with exit value:
{{case_clause,{badrpc,nodedown}},[{vmq_reg,'-register_subscriber/4-fun-0-',2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,182}]},{vmq_reg,block_until_migrated,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,249}]},{vmq_reg,register_subscriber,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,194}]},{vmq_reg_sync_action,'-init/1-fun-0-',1,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg_sync_action.erl"},{line,52}]}]}
2018-01-23 13:05:53 =CRASH REPORT====
  crasher:
    initial call: vmq_ranch:init/5
    pid: <0.15649.1>
    registered_name: []
    exception error: {{case_clause,{{case_clause,{badrpc,nodedown}},[{vmq_reg,'-register_subscriber/4-fun-0-',2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,182}]},{vmq_reg,block_until_migrated,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,249}]},{vmq_reg,register_subscriber,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,194}]},{vmq_reg_sync_action,'-init/1-fun-0-',1,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg_sync_action.erl"},{line,52}]}]}},[{vmq_mqtt_fsm,check_user,2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,552}]},{vmq_mqtt_fsm,in,3,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,173}]},{vmq_mqtt_fsm,data_in,3,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_mqtt_fsm.erl"},{line,141}]},{vmq_ranch,handle_message,2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_ranch.erl"},{line,163}]},{vmq_ranch,loop_,1,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_ranch.erl"},{line,106}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [<0.372.0>,<0.371.0>,ranch_sup,<0.71.0>]
    messages: []
    links: [<0.372.0>,#Port<0.17274>]
    dictionary: [{socket_open,<<>>},{max_msg_size,0},{rand_seed,{#{max => 288230376151711743,next => #Fun<rand.8.41921595>,type => exsplus,uniform => #Fun<rand.9.41921595>,uniform_n => #Fun<rand.10.41921595>},[285248026896301291|75704462735609890]}},{bytes_received,<<>>},{mqtt_connect_received,<<>>}]
    trap_exit: true
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 718
  neighbours:
2018-01-23 13:05:53 =ERROR REPORT====
Ranch listener {{0,0,0,0},5000} had connection process started with vmq_ranch:start_link/4 at <0.15649.1> exit with reason: {case_clause,{{case_clause,{badrpc,nodedown}},[{vmq_reg,'-register_subscriber/4-fun-0-',2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,182}]},{vmq_reg,block_until_migrated,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,249}]},{vmq_reg,register_subscriber,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,194}]},{vmq_reg_sync_action,'-init/1-fun-0-',1,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg_sync_action.erl"},{line,52}]}]}}```

Memory was fine...
tcp connections were set to 5000 now...
Nothign chnaged. Suddenly stopped working properly. Some requests were being catered to others being rejected. ALl nodes were up and running and custer status was true for all instances.

**Comment by ranvijayj** [comment]
```018-01-23 13:04:22.346 [error] <0.4508.2> CRASH REPORT Process <0.4508.2> with 0 neighbours crashed with reason: no case clause matching {{case_clause,{badrpc,nodedown}},[{vmq_reg,'-register_subscriber/4-fun-0-',2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,182}]},{vmq_reg,block_until_migrated,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,249}]},{vmq_reg,register_subscriber,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,194}]},{vmq_reg_sync_action,'-init/1-fun-0-',1,[{file,"/opt/vernemq/distdi..."},...]}]} in vmq_mqtt_fsm:check_user/2 line 552
2018-01-23 13:04:22.347 [error] <0.4459.2> Ranch listener {{0,0,0,0},5000} terminated with reason: {case_clause,{{case_clause,{badrpc,nodedown}},[{vmq_reg,'-register_subscriber/4-fun-0-',2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,182}]},{vmq_reg,block_until_migrated,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,249}]},{vmq_reg,register_subscriber,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,194}]},{vmq_reg_sync_action,'-init/1-fun-0-',1,[{file,"/opt/v..."},...]}]}}
2018-01-23 13:04:22.347 [error] <0.4508.2> Ranch listener {{0,0,0,0},5000} terminated with reason: {case_clause,{{case_clause,{badrpc,nodedown}},[{vmq_reg,'-register_subscriber/4-fun-0-',2,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,182}]},{vmq_reg,block_until_migrated,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,249}]},{vmq_reg,register_subscriber,4,[{file,"/opt/vernemq/distdir/1.2.1/_build/default/lib/vmq_server/src/vmq_reg.erl"},{line,194}]},{vmq_reg_sync_action,'-init/1-fun-0-',1,[{file,"/opt/v..."},...]}]}}
2018-01-23 13:04:25.956 [error] <0.551.2> CRASH REPORT Process <0.551.2> with 0 neighbours crashed with reason: bad argument in call to ets:lookup(vmq_trie, {trie_edge,{[],root},<<"#">>}) in vmq_reg_trie:'trie_match_#'/2 line 355
2018-01-23 13:04:25.957 [error] <0.551.2> Ranch listener {{0,0,0,0},1883} terminated with reason: badarg
2018-01-23 13:04:25.957 [error] <0.4599.2> CRASH REPORT Process <0.4599.2> with 0 neighbours crashed with reason: bad argument in call to ets:lookup(vmq_retain_srv, {[],[<<"user">>,<<"messages">>,<<"Njoksi3G753W7M0SfXZzTSSOQrBUcdEGtDGLLjwQ">>]}) in vmq_retain_srv:match_fold/4 line 91
2018-01-23 13:04:25.957 [error] <0.4599.2> Ranch listener {{0,0,0,0},5000} terminated with reason: badarg
2018-01-23 13:04:26.104 [error] <0.4647.2> CRASH REPORT Process <0.4647.2> with 0 neighbours crashed with reason: bad argument in call to ets:lookup(vmq_config_cache, ignore_db_config) in vmq_config:get_env/3 line 79```

**Comment by larshesel** [comment]
Hi @ranvijayj the `{badrpc, nodedown}` errors indicates that likely a node isn't reachable. We made the check if a remote node is available stronger in VerneMQ 1.2.3 as earlier version could show nodes as available although only the control channel but no data channel was established.

**Comment by dhaval1983** [comment]
I am getting the same error

Can someone please help
I am getting this error in log

ail -f error.log
2018-05-31 02:57:48.576 [error] <0.9249.0> CRASH REPORT Process <0.9249.0> with 0 neighbours crashed with reason: bad argument in call to ets:lookup(vmq_status, ready) in vmq_cluster:is_ready/0 line 68
2018-05-31 02:57:48.577 [error] <0.9249.0> Ranch listener {{0,0,0,0},2052} terminated with reason: badarg
2018-05-31 02:57:48.969 [error] <0.9250.0> CRASH REPORT Process <0.9250.0> with 0 neighbours crashed with reason: bad argument in call to ets:lookup(vmq_config_cache, ignore_db_config) in vmq_config:get_env/3 line 79
2018-05-31 02:57:48.969 [error] <0.9250.0> Ranch listener {{0,0,0,0},2052} terminated with reason: badarg
2018-05-31 02:57:49.083 [error] <0.9251.0> CRASH REPORT Process <0.9251.0> with 0 neighbours crashed with reason: bad argument in call to ets:lookup(vmq_config_cache, ignore_db_config) in vmq_config:get_env/3 line 79
2018-05-31 02:57:49.083 [error] <0.9251.0> Ranch listener {{0,0,0,0},2052} terminated with reason: badarg
2018-05-31 02:57:58.389 [error] <0.387.0> Failed to start Ranch listener {{0,0,0,0},8443} in ranch_ssl:listen([{ip,{0,0,0,0}},{port,8443},{nodelay,true},{linger,{true,0}},{send_timeout,30000},{send_timeout_close,true},{cacertfile,"/etc/vernemq/gd_bundle_domain.com.crt"},{certfile,"/etc/vernemq/domain.com.crt"},{keyfile,"/etc/vernemq/domain.com.key"},{ciphers,["ECDHE-ECDSA-AES256-GCM-SHA384","ECDHE-RSA-AES256-GCM-SHA384","ECDHE-ECDSA-AES256-SHA384","ECDHE-RSA-AES256-SHA384","ECDHE-ECDSA-DES-CBC3-SHA","ECDH-ECDSA-AES256-GCM-SHA384","ECDH-RSA-AES256-GCM-SHA384","ECDH-ECDSA-AES256-SHA384","ECDH-RSA-AES256-SHA384","DHE-DSS-AES256-GCM-SHA384","DHE-DSS-AES256-SHA256","AES256-GCM-SHA384","AES256-SHA256","ECDHE-ECDSA-AES128-GCM-SHA256","ECDHE-RSA-AES128-GCM-SHA256","ECDHE-ECDSA-AES128-SHA256","ECDHE-RSA-AES128-SHA256","ECDH-ECDSA-AES128-GCM-SHA256","ECDH-RSA-AES128-GCM-SHA256","ECDH-ECDSA-AES128-SHA256","ECDH-RSA-AES128-SHA256","DHE-DSS-AES128-GCM-SHA256","DHE-DSS-AES128-SHA256","AES128-GCM-SHA256","AES128-SHA256","ECDHE-ECDSA-AES256-SHA","ECDHE-RSA-AES256-SHA","DHE-DSS-AES256-SHA","ECDH-ECDSA-AES256-SHA","ECDH-RSA-AES256-SHA","AES256-SHA","ECDHE-ECDSA-AES128-SHA","ECDHE-RSA-AES128-SHA","DHE-DSS-AES128-SHA","ECDH-ECDSA-AES128-SHA","ECDH-RSA-AES128-SHA","AES128-SHA"]},{fail_if_no_peer_cert,false},{verify,verify_none},{verify_fun,{#Fun<vmq_ssl.0.123873112>,no_crl}},{depth,1},{versions,['tlsv1.2']}]) for reason eaddrinuse (address already in use)
2018-05-31 02:57:58.389 [error] <0.387.0> CRASH REPORT Process <0.387.0> with 0 neighbours exited with reason: {listen_error,{{0,0,0,0},8443},eaddrinuse} in gen_server:init_it/6 line 352
2018-05-31 02:57:58.389 [error] <0.245.0>@vmq_ranch_config:reconfigure_listeners_for_type:187 can't reconfigure mqttwss listener({0,0,0,0}, 8443) with Options [{max_connections,10000},{nr_of_acceptors,10},{mountpoint,"on"},{cafile,"/etc/vernemq/gd_bundle_domain.com.crt"},{depth,1},{certfile,"/etc/vernemq/domain.com.crt"},{keyfile,"/etc/vernemq/domain.com.key"},{require_certificate,false},{tls_version,'tlsv1.2'},{use_identity_as_username,false}] due to {{shutdown,{failed_to_start_child,ranch_acceptors_sup,{listen_error,{{0,0,0,0},8443},eaddrinuse}}},{child,undefined,{ranch_listener_sup,{{0,0,0,0},8443}},{ranch_listener_sup,start_link,[{{0,0,0,0},8443},10,ranch_ssl,[{ip,{0,0,0,0}},{port,8443},{nodelay,true},{linger,{true,0}},{send_timeout,30000},{send_timeout_close,true},{cacertfile,"/etc/vernemq/gd_bundle_domain.com.crt"},{certfile,"/etc/vernemq/domain.com.crt"},{keyfile,"/etc/vernemq/domain.com.key"},{ciphers,["ECDHE-ECDSA-AES256-GCM-SHA384","ECDHE-RSA-AES256-GCM-SHA384","ECDHE-ECDSA-AES256-SHA384","ECDHE-RSA-AES256-SHA384","ECDHE-ECDSA-DES-CBC3-SHA","ECDH-ECDSA-AES256-GCM-SHA384","ECDH-RSA-AES256-GCM-SHA384","ECDH-ECDSA-AES256-SHA384","ECDH-RSA-AES256-SHA384","DHE-DSS-AES256-GCM-SHA384","DHE-DSS-AES256-SHA256","AES256-GCM-SHA384","AES256-SHA256","ECDHE-ECDSA-AES128-GCM-SHA256","ECDHE-RSA-AES128-GCM-SHA256","ECDHE-ECDSA-AES128-SHA256","ECDHE-RSA-AES128-SHA256","ECDH-ECDSA-AES128-GCM-SHA256","ECDH-RSA-AES128-GCM-SHA256","ECDH-ECDSA-AES128-SHA256","ECDH-RSA-AES128-SHA256","DHE-DSS-AES128-GCM-SHA256","DHE-DSS-AES128-SHA256","AES128-GCM-SHA256","AES128-SHA256","ECDHE-ECDSA-AES256-SHA","ECDHE-RSA-AES256-SHA","DHE-DSS-AES256-SHA","ECDH-ECDSA-AES256-SHA","ECDH-RSA-AES256-SHA","AES256-SHA","ECDHE-ECDSA-AES128-SHA","ECDHE-RSA-AES128-SHA","DHE-DSS-AES128-SHA","ECDH-ECDSA-AES128-SHA","ECDH-RSA-AES128-SHA","AES128-SHA"]},{fail_if_no_peer_cert,false},{verify,verify_none},{verify_fun,{#Fun<vmq_ssl.0.123873112>,no_crl}},{depth,1},{versions,['tlsv1.2']}],cowboy_protocol,[{env,[{dispatch,[{'_',[],[{[<<"mqtt">>],[],vmq_websocket,[{mountpoint,"on"},{use_identity_as_username,false}]}]}]}]}]]},permanent,infinity,supervisor,[ranch_listener_sup]}}
2018-05-31 02:57:58.389 [error] <0.385.0> Supervisor {<0.385.0>,ranch_listener_sup} had child ranch_acceptors_sup started with ranch_acceptors_sup:start_link({{0,0,0,0},8443}, 10, ranch_ssl, [{ip,{0,0,0,0}},{port,8443},{nodelay,true},{linger,{true,0}},{send_timeout,30000},{send_timeout_close,...},...]) at undefined exit with reason {listen_error,{{0,0,0,0},8443},eaddrinuse} in context start_error

Can someone please help
